### ISSUE TITLE
Concurrent limited requests are not sent immediately

### ISSUE THREAD

**Issue by Briareos** [guzzle/guzzle#1125]
So this is a tricky one that's been following us since we started using Guzzle (v4); we found a workaround in Guzzle 5 (in RingPHP); but the issue returned in Guzzle 6.

The issue lies with `CurlMultiHandler`. When a request in created in a `complete` or `error` event (and now in middleware), it is added to the curl multi handle via `curl_multi_add_handle`. That's all good, but the HTTP request won't be sent until we call `curl_multi_exec`. The problem is, `curl_multi_exec` is called in `tick()` method **after** `curl_multi_select`, meaning that the handler will wait for any already sent request to complete before sending the request that was just added to the pool.

In Guzzle 4 I created a custom `CurlMultiHandler` that only added 3 lines to the original handler. I tried to get them into Guzzle 5, but it wasn't the most efficient method: https://github.com/guzzle/RingPHP/pull/18
After that I discovered that I can set the request's `future` option to `true`, and I would get exactly what I wanted, an additional call to `curl_multi_exec` exactly when I needed it: https://github.com/guzzle/RingPHP/blob/master/src/Client/CurlMultiHandler.php#L162

In Guzzle 6 I see this issue still persists. Here at ManageWP [we use Guzzle extensively](https://managewp.com/developing-backend-for-managewp-orion), so we forked our own `server.js` to support dynamic request duration and wrote tests for this exact case (every milisecond matters to us). For now I ended up creating a custom multi handler that calls additional `curl_multi_exec` if needed (kinda like RingPHP). @mtdowling, if I have your green light, I can port those tests and merge the `server.js` that supports dynamic request duration. With that said, I still don't know a better way to solve this issue. Maybe set a flag after the first call to `curl_multi_exec`, so any new requests (in `addRequest`) are sent immediately.

Let me know if I missed something or wasn't clear enough.

**Comment by mtdowling** [comment]
I would not want to always call `curl_multi_exec` again after processing the completed messages as that would significantly increase CPU utilization. However, maybe there's a way to determine if new requests were added to the multi handle while processing messages, and if so, call `curl_multi_exec`.

**Comment by Briareos** [comment]
Here's what we're using now:

In `tick`:
```php
// ...
// Step through the task queue which may add additional requests.
P\queue()->run();

if ($this->active &&
    curl_multi_select($this->_mh, $this->selectTimeout) === -1
) {
    // Perform a usleep if a select returns -1.
    // See: https://bugs.php.net/bug.php?id=61141
    usleep(250);
}
while (curl_multi_exec($this->_mh, $this->active) === CURLM_CALL_MULTI_PERFORM);

$this->running = true;

$this->processMessages();

if (!$this->active && !$this->delays && P\queue()->isEmpty()) {
    $this->running = false;
}
```

I've added `$this->running` just above and below `$this->processMessages();`.

Also, in `addRequest`:
```php
// ...
curl_multi_add_handle($this->_mh, $easy->handle);
if ($this->running) {
    while (curl_multi_exec($this->_mh, $this->active) === CURLM_CALL_MULTI_PERFORM);
    // If a request was completed between the last curl_multi_exec and this one,
    // we should process it right away; otherwise it will get processed only after
    // the next curl_multi_select call.
    $this->processMessages();
}
```

Does something like that make sense?

**Comment by mtdowling** [comment]
We don't want to process messages when adding requests as that can have unintended side effects (e.g., resolving requests and fulfilling promises before the stack unwinds in the promise trampoline causing recursion).

Instead of calling `curl_multi_exec` for each individual request added during a call to processMessages(), perhaps we can use a single extra `curl_multi_exec` if we detect that a handle was added while processing messages. This could be done using a simple `$state` flag on the class:

```php
    private function processMessages()
    {
        $this->state = self::STATE_PROCESSING;
        while ($done = curl_multi_info_read($this->_mh)) {
            $id = (int) $done['handle'];
            curl_multi_remove_handle($this->_mh, $done['handle']);
            if (!isset($this->handles[$id])) {
                // Probably was cancelled.
                continue;
            }
            $entry = $this->handles[$id];
            unset($this->handles[$id], $this->delays[$id]);
            $entry['easy']->errno = $done['result'];
            $entry['deferred']->resolve(
                CurlFactory::finish(
                    $this,
                    $entry['easy'],
                    $this->factory
                )
            );
        }

        if ($this->state === self::STATE_NEEDS_EXEC) {
            $this->state = self::STATE_DEFAULT;
            return true;
        }
        $this->state = self::STATE_DEFAULT;
        return false;
    }
```

```php
    private function addRequest(array $entry)
    {
        $easy = $entry['easy'];
        $id = (int) $easy->handle;
        $this->handles[$id] = $entry;
        if (empty($easy->options['delay'])) {
            curl_multi_add_handle($this->_mh, $easy->handle);
            if ($this->state === self::STATE_PROCESSING) {
                $this->state === self::STATE_NEEDS_EXEC;
            }
        } else {
            $this->delays[$id] = microtime(true) + ($easy->options['delay'] / 1000);
        }
    }
```

```php
    /**
     * Ticks the curl event loop.
     */
    public function tick()
    {
        // Add any delayed handles if needed.
        if ($this->delays) {
            $currentTime = microtime(true);
            foreach ($this->delays as $id => $delay) {
                if ($currentTime >= $delay) {
                    unset($this->delays[$id]);
                    curl_multi_add_handle(
                        $this->_mh,
                        $this->handles[$id]['easy']->handle
                    );
                }
            }
        }

        // Step through the task queue which may add additional requests.
        P\queue()->run();

        if ($this->active &&
            curl_multi_select($this->_mh, $this->selectTimeout) === -1
        ) {
            // Perform a usleep if a select returns -1.
            // See: https://bugs.php.net/bug.php?id=61141
            usleep(250);
        }

        do {
            while (curl_multi_exec($this->_mh, $this->active) === CURLM_CALL_MULTI_PERFORM);
        } while ($this->processMessages());
    }
```

Something like the above could work. What do you think?

**Comment by Briareos** [comment]
@mtdowling That makes sense, and it's also better performance-wise. Although, we have a special use case (or maybe it's not so special) that this solution doesn't fully resolve.

We have a throttling middleware that throttles request to a single server by intercepting a request with a promise:
```php
$promise = new P\Promise([P\queue(), 'run']);
```
, so the next request we send is added during
```php
// Step through the task queue which may add additional requests.
P\queue()->run();
```
inside the `tick()` method, and requests that are added there do not trigger `curl_multi_exec()` in time and are not sent immediately.

If we were to call `curl_multi_exec()` and process messages right after `P\queue()->run();`, there could be new messages in `P\queue()->run();`, so we would have to put it inside a loop too.

This is what I ended up with:
```php
    /**
     * Ticks the curl event loop.
     */
    public function tick()
    {
        // Add any delayed handles if needed.
        if ($this->delays) {
            $currentTime = microtime(true);
            foreach ($this->delays as $id => $delay) {
                if ($currentTime >= $delay) {
                    unset($this->delays[$id]);
                    curl_multi_add_handle(
                        $this->_mh,
                        $this->handles[$id]['easy']->handle
                    );
                }
            }
        }

        do {
            // Step through the task queue which may add additional requests.
            P\queue()->run();

            $queueClean = true;

            if ($this->state === self::STATE_NEEDS_EXEC) {
                $queueClean = false;
                do {
                    while (curl_multi_exec($this->_mh, $this->active) === CURLM_CALL_MULTI_PERFORM);
                } while ($this->processMessages());
            }
        } while (!$queueClean);

        if ($this->active &&
            curl_multi_select($this->_mh, $this->selectTimeout) === -1
        ) {
            // Perform a usleep if a select returns -1.
            // See: https://bugs.php.net/bug.php?id=61141
            usleep(250);
        }

        do {
            while (curl_multi_exec($this->_mh, $this->active) === CURLM_CALL_MULTI_PERFORM);
        } while ($this->processMessages());
    }
```

```php
    private function addRequest(array $entry)
    {
        $easy = $entry['easy'];
        $id = (int) $easy->handle;
        $this->handles[$id] = $entry;
        if (empty($easy->options['delay'])) {
            curl_multi_add_handle($this->_mh, $easy->handle);
            $this->state = self::STATE_NEEDS_EXEC;
        } else {
            $this->delays[$id] = microtime(true) + ($easy->options['delay'] / 1000);
        }
    }
```

It might contain some redundant code, but it passes all our tests. The only different behavior is, the first call to `tick()` is blocking until a first response is received by cURL, and it kinda conflicts with what I said in https://github.com/guzzle/guzzle/issues/1127#issuecomment-112437287.
Thoughts?

**Comment by mtdowling** [comment]
I think this is adding up to a lot of changes to address more or less an edge case, and I'd prefer to not place all of that additional logic in the main curl multi loop if possible.

Maybe there's a better way that does what we both want but in fewer code changes that don't really affect the performance of the main curl loop. Maybe it would be easier to asses this in a pull request.
